src/GraphQL/Server.hs

```diff
@@ -180,27 +196,37 @@ instance forall v. ReadValue v => ReadValue (Maybe v) where
 -- the name matches the query. Note that the name is *not* in monad m,
 -- but the value is. This is necessary so we can skip execution if the
 -- name doesn't match.
-data NamedFieldExecutor m = NamedFieldExecutor AST.Name (m GValue.Value)
-
-executeNamedField :: Monad m => AST.Field -> NamedFieldExecutor m -> m (Maybe GValue.ObjectField)
-executeNamedField (AST.Field alias name _ _ _) (NamedFieldExecutor k mValue)
-  | name == k = do
-      value <- mValue
-      let name' = fromMaybe name alias
-      pure (Just (GValue.ObjectField name' value))
-  | otherwise = pure Nothing
-
--- Deconstruct object type signature and handler value at the same
--- time and run type-directed code for each field.
-resolveField :: forall a (m :: Type -> Type). BuildFieldResolver m a => m GValue.ObjectField -> FieldHandler m a -> AST.Selection -> m GValue.ObjectField
-resolveField fallback lh selection@(AST.SelectionField field) = do
-  let fieldExecutor = buildFieldResolver @m @a lh selection
-  case fieldExecutor of
-    Left err -> throwM err
-    Right executor -> do
-      objectField <- executeNamedField field executor
-      maybe fallback pure objectField
-resolveField _ _ f = queryError ("Unexpected Selection value. Is the query normalized?: " <> show f)
+data NamedValueResolver m = NamedValueResolver AST.Name (m Result)
+
+
+
+-- Iterate through handlers (zipped together with their type
+-- definition) and execute handler if the name matches.
+--
+-- TODO: tuple return not great (but either is wrong, too because it discards errors.
+type ResolveFieldResult = ([QueryError], Maybe GValue.ObjectField)
```

Comment:
I've changed this to the parametrized result but I think it gives a false sense of generalization - `ObjectField` is so different from `Value` that we can't even have a generic aggregator. IMO it'd be better to stick with some module-local `ResolveFieldResult` type to avoid giving the impression that it's anything like result. What do you think?